### DISCUSSION TITLE
RP2040

### DISCUSSION THREAD

**Discussion by JohnK1987** [mbed-ce/mbed-os#106] [Ideas]
Hello,

we already discussed (Mbed's Forum) about RP2040 and unfinished PR - https://github.com/ARMmbed/mbed-os/pull/14488
I already check that and after few corrections it is compilable, but cannot verify that because a standart output is .hex or .bin and this not enough and a format .uf2 is required.
[I found](https://shawnhymel.com/2096/how-to-set-up-raspberry-pi-pico-c-c-toolchain-on-windows-with-vs-code/#Install_MinGW-w64_GCC_Tools) all necessary things are already in PicoSDK only MinGW-w64_GCC_Tools must be installed, but I have no idea how to set this up  together with MbedCE in VS Code.

Do you have any ideas guys? @JojoS62 @multiplemonomials 

BR, Jan

**Comment by JojoS62** [comment]
It is interesting, although the official Arduino Core uses Mbed and so you can use Mbed classes.
For a good integration into Mbed CE it needs a new upload method for uf2. The SDK tool is running fine under Linux or in WSL, but I have also struggled with this in Windows. It could be easier when there is a python module for handling uf2.
Another tool is BMP, black magic probe and has an integrated gdbserver and good cortex-debug support. Afaik this works now also with RP2040.

**Comment by JohnK1987** [comment]
When you compare [ArduinoCore-mbed/targets/TARGET_RASPBERRYPI/TARGET_RP2040 ](https://github.com/arduino/ArduinoCore-mbed/tree/master/cores/arduino/mbed/targets/TARGET_RASPBERRYPI/TARGET_RP2040) with the content of original PR it is little bit different (check the target folder in the tree) - https://github.dev/ARMmbed/mbed-os/pull/14488. Also latest ArduinoCore-mbed probably uses newer version of PicoSDK. 
Moreover https://github.com/arduino/ArduinoCore-mbed/tree/master/patches

I do not have [BMP](https://black-magic.org/) but PicoProbe seems be also promising. Ofc yeah, it is just for RP.
PicoProbe:
- openOCD - https://github.com/openocd-org/openocd/releases/tag/v0.12.0-rc1
- pyOCD - https://pyocd.io/docs/debug_probes.html

BR, Jan

**Reply by JohnK1987** [reply]
According to page 29 of  [getting-started-with-pico](https://datasheets.raspberrypi.com/pico/getting-started-with-pico.pdf) seems there is a problem with usage of the ninja
> Depending on your local setup you may not need to set the CMake generator manually to "Unix Makefiles". However
if you do not do so in some cases Visual Studio will default to ninja rather than make and the build might fail as GCC
outputs dependency-information in a slightly-incorrect format that ninja can’t understand.
If you do find yourself having to configure this variable manually it is also possible that you may need to point Visual
Studio at the CMake Tools extension explicitly by adding an additional line to your settings.json file,
{
  "cmake.environment": {
  "PICO_SDK_PATH": "../../pico-sdk"
  },
  "C_Cpp.default.configurationProvider": "ms-vscode.cmake-tools"
}

@multiplemonomials is this a probem for us? Thank you.

BR, Jan

**Reply by multiplemonomials** [reply]
hmm, I doubt it, unless you have to use a toolchain other than the the regular ARM none eabi GCC to compile for it.

**Reply by JohnK1987** [reply]
Ok, I am able to compile and upload Mbed blinky to Raspberry Pi Pico via the picoprobe and pyOCD for now! :) As soon as I add printf, then the program stucks. I will continue and we will see.

BR, Jan

**Reply by multiplemonomials** [reply]
Oh nice!  Share your branch!  I actually bought a pico a while back and am planning to do a debugger mod on it

**Reply by JohnK1987** [reply]
It is based on links mentioned above
* Here is link to the [branch](https://github.com/JohnK1987/mbed-os/tree/RP2040)
* Here is a link for information collection and preparation of [future MCU page](https://github.com/mbed-ce/mbed-os/wiki/%5BIn-preparation%5D-MCU-Info-Page:-RP2040)
BR, Jan

**Reply by JojoS62** [reply]
thanks for touching the RP2040, currently I have also connected a RPi pico board. It is also a very powerful MCU, I have connected a TFT and it is very fast with 62,5 MHz SPI and also running lvgl graphics. Yesterday, I got a new board, tufty2040 from pimoroni.com. A cute board with good quality TFT.
I have used platformIO and started with the official Mbed Arduino core. It works, but is not well maintained. There are a lot of issues and building throws zillions of warnings. So the earlephilhower core is more popular and e.g. has faster SPI. Also reading µs has to pass several layers and takes many microseconds (!) to read the current value. A good chance to improve things in Mbed-CE.
The picoprobe code is public, but it looks like it is a simple USB-UART interface and it is used with openOCD in some bitbanging mode: https://github.com/raspberrypi/picoprobe
A possible alternative is Black Magic Probe, my favorite probe beside STLinkV3. I haven't tried the newer release of BMP, I hope it supports RP2040 now. For the dual core the probe has to support multidrop mode to address the different cores.
Multicore support is also easy in the  earlephilhower core, one more point why it is more popular.

**Reply by JohnK1987** [reply]
My goal at the moment is to make the RP2040 usable in MbedCE with passed Greentea test (minor issues are acceptable).

I am using Picoprobe with pyOCD because it was very easy to setup and I found some references for that. Current OpenOCD (0.11) does not contain picoprobe and I do not want download another version of OpenOCD (fork of officiall one) from Rapsberry repo, but I hope incomming new version (0.12) of officiall OpenOCD will contain that.

About multicore. As I know Mbed is based on ARM RTX and the RTX is not able to run on multiple cores from one instace of the OS. That is possible to do with freeRTOS. (That lead me to interesting question... Mbed based on freeRTOS, ehm nwm :) ). I am not so far and I am not sure what about two separete projects, project for each core, but I read some discussions and according to experiences of Arduino users the Pico is not able to run any Mbed code that contains core_critical_section..._ on the Core1 from program for Core 0. It is possible to do that when the code will be based just on PICO SDK's functions.

**Reply by multiplemonomials** [reply]
> Picoprobe

I would think that the best cheap debugger option for the RP2040 is the [PitayaLink](https://makerdiary.com/products/pitaya-link), because unlike the picoprobe it supports SWD and UART at the same time.  Like the picoprobe, it's usable with OpenOCD and PyOCD as a generic CMSIS-DAP interface.

> Mbed based on freeRTOS

That's actually something I've been considering as well.  Theoretically it is possible, because FreeRTOS has an API layer that makes it conform to the CMSIS-RTOS standard.  It would take a bit of work though because the configuration would need to be set up, and some of the RTX-specific stuff in Mbed would need to be rewritten.

The way I see it, the biggest benefit would be better debugger support, since OpenOCD and ST-Link know how to scan FreeRTOS threads.  But there are probably some other benefits since the FreeRTOS codebase is likely better tested and more well-known in general.

**Reply by JohnK1987** [reply]
Yeah, Pitaya seems good, I will buy one maybe in the future.

About freeRTOS it is a question, 
However how Jojo wrote above about the popularity of [Earlephilhower-Arduino-pico](https://github.com/earlephilhower/arduino-pico), this also uses also freeRTOS and I believe it is because of multi core support.
Maybe we could make separate discussion and analysis complication and expected result of this before a decision.

With Pico nothing new for now. For some reason the Pico does not want to send anything to UART, and with BufferedSerial (not with unbuffered) it also freezes  after 17 loops.

**Reply by JohnK1987** [reply]
So, BufferedSerial is not working for some reason and I still do not know why, but I found it was never working probably, because STDIO is redirected to UnbuffereSerial that works gread... -  `"platform.stdio-buffered-serial": false,`

How I am finding more and more small things around RP2040 I more and more feel why the original PR was never Merged... a mess.

**Reply by JohnK1987** [reply]
Currently I am struggling with BufferedSerial. I found program stops in this [loop](https://github.com/mbed-ce/mbed-os/blob/af6d50aefe923349c32c8a1bd6dd0786ffe66fd7/drivers/source/BufferedSerial.cpp#L302-L304) where is called a method `SerialBase::writeable()` that returns `false` sometimes for some reason (usually after first two chars)  (I also tried that directly with HALs from PICO_SDK and the behavior was same). 
When this happen then should come [TX_IRQ](https://github.com/mbed-ce/mbed-os/blob/af6d50aefe923349c32c8a1bd6dd0786ffe66fd7/drivers/source/BufferedSerial.cpp#L340-L351) on the scene and solve that, however nothing will happen, so the loop continues forever.
So if I understand that correctly then [here](https://github.com/JohnK1987/mbed-os/blob/ed6a06023ba02d6bdb801f9ab744c66ba5a49292/targets/TARGET_RASPBERRYPI/TARGET_RP2040/serial_api.c#L102-L134) is not TxIrq solved correctly.

Another issue where I cannot identify its source is 0X00 on TX directly after power-up and before a first UART0 init. And more strange behavior is that how I run UnbufferedSerial echo example then the serial monitor is continuously spammed with these 0x00.

@multiplemonomials it seems the UART of Picoprobe is possible to use during debugging without any visible restrictions right now.

BR, Jan

**Reply by multiplemonomials** [reply]
Hmm, I'm looking into how we can get hex files into a compatible uf2 file.  This article is pretty interesting: https://makecode.com/blog/one-chip-to-flash-them-all

**Reply by JohnK1987** [reply]
It is not necessary for now, I think. How I wrote above currently the Picoprobe with pyOCD is working. From my point of view the flash and debug functionality > uf2.
However how I remember (also description from last post) the main problem seems to be issue with BufferedSrial API especialy with TX IRQ.

I do nothing on this at the moment. I am working on STM32H5 during free time because I am working on my house so I have very limited time.

BR, Jan

**Reply by multiplemonomials** [reply]
All right, I was able to get the RP2040 talking really easily with OpenOCD and a Pitaya-Link debug probe!
![image](https://github.com/mbed-ce/mbed-os/assets/2099358/70cf9dbd-4b8c-4863-a469-2ab36e1818d4)

Now let's see if I can flash it...

**Reply by multiplemonomials** [reply]
Alright, I got to the point where I can run the test suite on it!
```
79% tests passed, 15 tests failed out of 71

Total Test time (real) = 1920.33 sec

The following tests did not run:
          8 - mbed-hal-crc (Skipped)
         11 - mbed-hal-lp-ticker (Skipped)
         13 - mbed-hal-mpu (Skipped)
         14 - mbed-hal-ospi (Skipped)
         16 - mbed-hal-qspi (Skipped)
         22 - mbed-hal-sleep (Skipped)
         23 - mbed-hal-sleep-manager (Skipped)
         24 - mbed-hal-sleep-manager-racecondition (Skipped)
         27 - mbed-hal-trng (Skipped)
         32 - mbed-hal-verify-arduino-pinmap (Skipped)
         43 - mbed-platform-crash-reporting (Skipped)
         53 - mbed-platform-stats-cpu (Skipped)
         54 - mbed-platform-stats-heap (Skipped)
         55 - mbed-platform-stats-sys (Skipped)
         71 - mbed-connectivity-nfc-eeprom (Skipped)

The following tests FAILED:
          2 - mbed-usb-device-basic (Failed)
          3 - mbed-usb-device-hid (Failed)
          4 - mbed-usb-device-msd (Failed)
          5 - mbed-usb-device-serial (Failed)
          6 - mbed-hal-common-tickers (Failed)
          7 - mbed-hal-common-tickers-freq (Failed)
         10 - mbed-hal-flash-functional-tests (Failed)
         15 - mbed-hal-pinmap (Failed)
         17 - mbed-hal-reset-reason (Failed)
         18 - mbed-hal-rtc (Failed)
         19 - mbed-hal-rtc-reset (Failed)
         20 - mbed-hal-rtc-time (Failed)
         25 - mbed-hal-stack-size-unification (Failed)
         29 - mbed-hal-watchdog (Failed)
         31 - mbed-hal-watchdog-timing (Failed)
```

Looks like USB is a bit of a dumpster fire at present, every test fails or hangs.  Also, the watchdog and RTC need help.  (the stack-size-unification test was due to a missing constant, I just fixed that one).  But all in all, not a bad start!  Note that this test run was using the default serial, not BufferedSerial

**Reply by multiplemonomials** [reply]
My branch: https://github.com/mbed-ce/mbed-os/tree/RP2040

**Reply by multiplemonomials** [reply]
Good news!  After a bit of debugging, I was able to fix the BufferedSerial issue.  Turns out that the problem was pretty simple -- no one had implemented the ability to set and use Tx IRQs in the serial_api driver.  The code was broken in such a way that it ended up delivering Rx IRQs when it meant to send Tx IRQs.  This, of course, caused BufferedSerial to not work as it would get stuck in an infinite ISR handler after sending the first few bytes.  I implemented the missing code and now it works -- check my branch above!

Also, I added Picotool upload method support and fleshed out the [wiki page](https://github.com/mbed-ce/mbed-os/wiki/%5BIn-preparation%5D-MCU-Info-Page:-RP2040) a lot.

**Reply by JohnK1987** [reply]
I know that was wrong but I was not able to find how to fix this. Good job.
I will try to fill some more info about Picoprobe into wiki page.

**Reply by JojoS62** [reply]
I've tried the RP2040 branch now also and first steps worked, a simple test program with EventQueue and Threads. Using the Picoprobe and OpenOCD under Windows.
There are a lot of warnings during compiling, signedness and missing return value, will check it again.
Great work!

**Comment by JojoS62** [comment]
that is interesting. There are already H7 multicore MCUs, how are they handled? Loading two independent projects?

PlatformIO comes with a modified version of OOCD for the pico, But you're right, it will be better to wait for an official new release.

**Reply by JohnK1987** [reply]
It seems like that - https://forums.mbed.com/t/mbed-os-new-features/11248/17

**Reply by JohnK1987** [reply]
@JojoS62 can you help me please? I am trying build BMP for ST-Link V2.1 according to [this](https://black-magic.org/hacking/hacking.html) failed

**Reply by JojoS62** [reply]
I can check tomorrow. Can you build under Linux or in WSL? I remember that the build process was not Windows friendly.

**Comment by JojoS62** [comment]
still have problems with my eyes and cannot stare for long time at the PC monitor. There is a Discord channel #blackmagic where you get good support: https://discord.gg/CFDxXKZW
The instructions for compiling are here: https://black-magic.org/hacking/hacking.html
But I don't know if they are up-to-date. There was a lot of revamping in the project in the past.
When you succeeded with flashing the probe, you should try first the manual use of gdb:
https://black-magic.org/usage/gdb-commands.html
Then create a debug configuration in VSC for cortex-debug, with servertype bmp. The only settings should be path to .elf file and the com port for the probe.

**Reply by JohnK1987** [reply]
Jojo you always create new conversation, because you select wrong Write window/line probably. Try to use window with Reply button not with Comment button :)
Oh, understand, then don't strain your eyes for stupid things. I will left this for some time, maybe forever because this is really pain, I do not understand this. I believe I made some mistakes somewhere, but... I do not want to know how solve every problem on internet :)

Just a summary.
I just wanted try the product BMP on ST-link. According to https://black-magic.org/hacking/hacking.html 
```
C:\Users\User\BMP\blackmagic>make PROBE_HOST=stlink ST_BOOTLOADER=1
! was unexpected at this time.
make: *** [all] Error 255
```
I did it under Windows, because I do not see any restriction for Linux ONLY!
According to your recommendation I installed WSL. Inside of WSL other things (Python, GNU...) what are already in my PC (waste of space). Different error, but result same.
```
root@DESKTOP-SBJJCIP:/home/johnnyk/Test/blackmagic# make PROBE_HOST=stlink ST_BOOTLOADER=1
  BUILD   lib/stm32/f1
make[2]: Nothing to be done for 'all'.
  BUILD   lib/stm32/f4
make[2]: Nothing to be done for 'all'.
  BUILD   lib/lm4f
make[2]: Nothing to be done for 'all'.
Load address 0x08004000 for original ST-LinkV2 Bootloader
  LD      blackmagic.elf
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: address 0x8020878 of blackmagic.elf section `.text' is not within region `rom'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: blackmagic.elf section `.ARM.exidx' will not fit in region `rom'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: address 0x8020878 of blackmagic.elf section `.text' is not within region `rom'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: region `rom' overflowed by 2464 bytes
Memory region         Used Size  Region Size  %age Used
             rom:      133536 B       128 KB    101.88%
             ram:       11604 B        20 KB     56.66%
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:120: blackmagic.elf] Error 1
make: *** [Makefile:17: all] Error 2
```

Then, I found this - https://github.com/blackmagic-debug/blackmagic/discussions/1368#discussioncomment-4805005
So I deleted some files (corresponding to non ST targets) from blackmagic\src\Makefile
```
Memory region         Used Size  Region Size  %age Used
             rom:      123144 B       128 KB     93.95%
             ram:       11920 B        20 KB     58.20%
  OBJCOPY blackmagic.bin
```
After hours I had binary (probably for a ST-Link) and I noticed I need another tool [stlink-tool](https://github.com/UweBonnes/stlink-tool/tree/stlinkv21), again need to be builded, again it is not working. This time because of _libusb.h: No such file or directory_. I tried to solve this (I already have it installed in WSL), but I am done.

Eh... they want 68 dolars for their probe (Official hardware), when they are not able to make simple instructions for their product? No, thank you, nothing for me :)

**Comment by JohnK1987** [comment]
@multiplemonomials I added another info for the picoprobe to [wiki page]( https://github.com/mbed-ce/mbed-os/wiki/MCU-Info-Page:-RP2040#raspberry-pi-picoprobe), hope it will be useful.

**Comment by JojoS62** [comment]
I have build the blinky for the tufty 2040 board from Pimoroni, it works. Used picotool upload in WSL2.
I had started a project using this target with PlatformIO, that was a little bit easier, but using the Earl Phil Power core. However, that is pretty fast with SPI DMA and PIO from the Bodmer/tft_eSPI Lib. So the next is to get this running also with Mbed.

**Reply by multiplemonomials** [reply]
Awesome!  Yeah it would pretty awesome if SPI DMA support could be a thing for RPi Pico!

**Reply by JohnK1987** [reply]
I have Arduino nano rp2040 connect on my table, but no space in calendar :)
